### PR TITLE
Whitelisting changes

### PR DESCRIPTION
**Pull request by PromoFaux**
Note : This will break the "ugly hack"  @jacobsalmela put in yesterday...

### CODE REVIEW

**Comment by dschaper** [comment]
Thanks!

**Comment by Mcat12** [comment]
By the way, this PR will rebase whitelist with the changes from master. (I don't know if I was the only one wondering why there were so many commits :) )

**Comment by Mcat12** [comment]
You should add in the commands list the quiet flag. It's implemented, but not shown with the other commands.

File: gravity.sh [hunk 1]
```diff
@@ -160,48 +159,35 @@ function gravity_Schwarzchild() {
...
-function gravity_pulsar() {
 
+function gravity_Blacklist(){
 	# Append blacklist entries if they exist
-	if [[ -r $blacklist ]];then
-        numberOf=$(cat $blacklist | sed '/^\s*$/d' | wc -l)
-        echo "** Blacklisting $numberOf domain(s)..."
-        cat $blacklist >> $piholeDir/$matter
-	fi
+	blacklist.sh -f -nr -q
```
Comment (Mcat12) [inline gravity.sh]: Do you need to add the domains you are blacklisting to this command?

**Comment by Mcat12** [comment]
Once this is merged, I can change up my white/blacklist editor to use the new features.

File: gravity.sh (hunk 1 above)
Comment (PromoFaux) [inline gravity.sh]: All handled in blacklist.sh!

**Comment by PromoFaux** [comment]
>You should add in the commands list the quiet flag. It's implemented, but not shown with the other commands.

Yeah, could do, I left it off of the list because it's only really for it being called from gravity. If a user is running it manually, they probably want some confirmation that the domains they added were blacklisted.

When run by gravity.sh, it a) makes sure the ad list domains are whitelisted, and b) just re-whitelists everything that's in whitelist.txt

**Comment by PromoFaux** [comment]
>  (I don't know if I was the only one wondering why there were so many commits :) )

There weren't when I originally set the pull request to master.. but since you suggested I merge with the whitelist branch,,,

File: advanced/Scripts/whitelist.sh [hunk 2]
```diff
@@ -7,83 +7,174 @@
 # the Free Software Foundation, either version 2 of the License, or
 # (at your option) any later version.
 
+if [[ $# = 0 ]]; then
+    echo "Immediately whitelists one or more domains in the hosts file"
+    echo " "
+    echo "Usage: whitelist.sh domain1 [domain2 ...]"
+    echo "  "
```
Comment (Mcat12) [inline advanced/Scripts/whitelist.sh]: Slight ocd on my part, but blacklist.sh has no empty lines between the help output.

**Comment by Mcat12** [comment]
No, it's good that the whitelist branch be updated (I think). We can make sure we haven't broken anything major.

File: gravity.sh (hunk 1 above)
Comment (Mcat12) [inline gravity.sh]: Nice. :smile:

File: advanced/Scripts/whitelist.sh (hunk 2 above)
Comment (PromoFaux) [inline advanced/Scripts/whitelist.sh]: Huh, thought i'd sorted that.. I think I might have got a bit confused with the merging earlier! Will sort it out in a bit

**Comment by PromoFaux** [comment]
There was some stuff committed into the whitelist.sh with a web hook labelled as a "dirty hack" so I just wrote my changes over it, as it probably would not have worked with my new changes anyway

**Comment by Mcat12** [comment]
That was @jacobsalmela trying to make a simple whitelister on the site, which would give the console output. It would redirect the console output to somewhere the script could get it and display on the site. The whitelister will soon be deprecated, if my editor is merged.

**Comment by Mcat12** [comment]
Hmm, I guess the whitelist branch already got the latest from master. You just were working with an earlier version, and now git is bringing your branch up to date. Ok, so if nothing is breaking now, it should stay safe. :smile:

File: gravity.sh (hunk 1 above)
Comment (PromoFaux) [inline gravity.sh]: Yeah, basically all that line says is "Run the blacklist script, force it to update the hosts file, don't reload dnsmasq, and do it quietly.

File: advanced/Scripts/whitelist.sh (hunk 2 above)
Comment (PromoFaux) [inline advanced/Scripts/whitelist.sh]: fixed in f9a2ca5

**Comment by dschaper** [comment]
What should we do if a user has a domain in both the blacklist and the whitelist? (Say they blacklist a domain, forget they did and a few months later try to whitelist the domain...) Just a hypothetical, not a comment on the current code. Testing out the code now and it's looking good. Thanks!

**Comment by PromoFaux** [comment]
Funny you should mention that, I've been giving it some thought this evening.

What I'll probably end up doing is giving both the whitelist.sh and blacklist.sh knowledge of each other's lists. 

e.g if you add a domain to the whitelist, it will be removed from the blacklist and vice versa. We should probably decide which one takes priority.

**Comment by dschaper** [comment]
I do a great impression of a forgetful user! (There is no truth to the rumor that the above scenario is based on reality in any way!)

I think whitelisting should be the winner, but that's just a biased coin flip, really.

**Comment by dschaper** [comment]
And now that I think about it some more, how do we handle removing from blacklist/whitelist. Use the same process and have whitelist remove the blacklist and blacklist remove the whitelist? Just throwing things out there. I've added domains with the blacklist script and for now I have to edit `gravity.list`  by hand or run `gravity.sh` again after cleaning the `blacklist.txt`. 

I guess with the improved web interface these situations will be uncommon, it will be easy to see what's what...

**Comment by PromoFaux** [comment]
To remove from the blacklist (and also remove it from gravity.list) run blacklist.sh -d <domain>

**Comment by dschaper** [comment]
Ah, thanks! (I'll document that in the Wiki when the PR merges.)

**Comment by PromoFaux** [comment]
I tested it as best I could earlier, but do let me know if you find any bugs, and I'll sort them out! Once it's merged, i'll start work on making the black and whitelists aware of each other.

**Comment by PromoFaux** [comment]
Oh also, try passing a non domain argument to either script! Quite proud of that bit..

e.g <code>blacklist.sh kjngkdsfnshkdfhn<code>

**Comment by dschaper** [comment]
Nice, lol... I'm still in a love/hate mode with regex.

**Comment by Mcat12** [comment]
I think the current gravity.sh has blacklist take priority (runs whitelist first, then blacklist). Are you planning on implementing a smart listing system as described above, or just giving one list priority?

**Comment by dschaper** [comment]
When I wrote `gravity_pulsar()` I didn't take into account any order, it's just happenstance that the order is like that... @jacobsalmela may have a preference though.

**Comment by jacobsalmela** [comment]
Wow guys, I was busy today but my inbox was blowing up with all the chatter.  I'll have to check this stuff and all the comments on the other issues/PRs.

Thanks for your hard work.

**Comment by jacobsalmela** [comment]
Impressive @PromoFaux!  

@Mcat12, should I wait to merge your changes on the Web UI?

**Comment by Mcat12** [comment]
Ya, I'm going to reinstall my pihole (was running off the dietpi version using apache) and implement the new whitelist and blacklist features. In the mean time, the web interface needs sudo for white/blacklist.sh, in order to restart the dns. I think this has to be done by giving www-data no password, no tty required sudo permissions on the scripts in the sudoers file.

**Comment by jacobsalmela** [comment]
Right, I was going to work on that.  

I'm going to focus on the bugs first, but go ahead and keep working so it's ready.   You are obviously better qualified for Web development then me!

**Comment by Mcat12** [comment]
lol, I guess I updated my pi too soon. I'm going to have to update again with all the changes committed.

**Comment by jacobsalmela** [comment]
Yeah sorry about that.; I was kind of on a spree and couldn't stop.

**Pull request merged**